### ISSUE TITLE
[Bug] openMSX fails to cross compile with MinGW with _clock_gettime link error [sf#499]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#893]
**Reported by manuelbi on 2013-09-19 20:01 UTC**
At some point this year, openMSX started failing to link with MinGW on Linux, using  mingw-w64-i686-dev 3.0~svn6246-1 on Debian.
Version is:

$ /usr/bin/i686-w64-mingw32-g++ --version
i686-w64-mingw32-g++ (GCC) 4.6.3

This is the link error:

Linking openmsx.exe...
derived/x86-mingw32-i686-3rd/obj/thread/Timer.o:Timer.cc:(.text+0x15): undefined reference to `_clock_gettime'
collect2: ld returned 1 exit status

Some relevant discussion:

jul 08 2013 23:12:02 <Quibus>   Timer.o indeed gives:
jul 08 2013 23:12:07 <Quibus>            U _clock_gettime
jul 08 2013 23:12:38 <Quibus>   so, how did it get in a heder file?
jul 08 2013 23:13:28 <Quibus>   that's pretty complicated
jul 08 2013 23:13:50 <Quibus>   #if HAVE_CLOCK_GETTIME
jul 08 2013 23:13:50 <Quibus>   #include <ctime>
jul 08 2013 23:13:50 <Quibus>   #endif
jul 08 2013 23:13:55 <Quibus>   so, it's getting it from ctime
jul 08 2013 23:16:28 <Quibus>   Did harlok also get this particular link error?
jul 08 2013 23:17:09 <mth>      dunno, I didn't see his full log
jul 08 2013 23:17:18 <mth>      check which lib defines the symbol
jul 08 2013 23:17:26 <mth>      it might just be a missing -l
jul 08 2013 23:17:29 <Quibus>   I don't even see a ctime header
jul 08 2013 23:17:39 <Quibus>   mth: note the underscore
jul 08 2013 23:17:52 <Quibus>   in openMSX source it's without underscore
jul 08 2013 23:18:11 <mth>      symbol name is not necessarily equal to function name
jul 08 2013 23:18:53 <mth>      the probe only checks the header for clock_gettime, it doesn't try to link
jul 08 2013 23:20:07 <mth>      'ctime' would be inside one of the GCC dirs, not in the main include dir, as it's a C++ header
jul 08 2013 23:20:41 <Quibus>   I don't know where I can find those for the mingw toolchain
jul 08 2013 23:22:45 <Quibus>   hmm
jul 08 2013 23:22:49 <Quibus>   000000e0 T _clock_gettime
jul 08 2013 23:22:57 <Quibus>   in libwinpthread_la-clock.o:
jul 08 2013 23:23:52 <Quibus>   /usr/i686-w64-mingw32/lib/libwinpthread-1.dll even
jul 08 2013 23:24:31 <Quibus>   my latest MinGW doc says:
jul 08 2013 23:24:44 <Quibus>   (in the changelog)
jul 08 2013 23:24:52 <Quibus>      - Build genpeimg, libwinpthreads (which replaces
jul 08 2013 23:24:52 <Quibus>         libpthreads-mingw-w64), libwinstorecompat and widl.
jul 08 2013 23:27:03 <Quibus>   so, it seems it needs to link to libwinpthread.dll.a or so
jul 08 2013 23:28:17 <Quibus>   so, I guess a link flag issue indeed
jul 08 2013 23:29:54 <Quibus>   perhaps this wasn't implemented in MinGW before, but now they added it to the header so the include test succeeds
jul 08 2013 23:30:31 <mth>      possible

[we find out that BiFi uses gcc 4.7.0 in his MinGW, and no problems there, although it's a local compile. We check the probe output.]

jul 09 2013 22:54:42 <Quibus>   mth: so in BiFi's MiNGW, there is no HAVE_CLOCK_GETTIME and in mine there is. Probably because they enabled that winpthreads library which contains that symbol
jul 09 2013 23:06:44 <Quibus>   (and we don't link against that lib)
jul 09 2013 23:07:41 <Quibus>   mth: so, what is a good way to solve it? We could link against the lib (but I don't know how compatible that is with other/older MiNGW versions). Or perhaps we can force HAVE_CLOCK_GETTIME to 0 if we're on MinGW
jul 10 2013 19:30:40 <mth>      Quibus: easiest is to link against winpthreads unconditionally, but you'll have to ask BiFi whether that works on older mingw installs
jul 10 2013 19:31:00 <mth>      second easiest is to do as wouterv said and test for the library presence
jul 10 2013 19:31:36 <mth>      forcing it to off is something we should only do as a last resort
jul 10 2013 21:11:24 <Quibus>   mth: if we'd just do a kind of link test, it would probably suffice, right?
jul 10 2013 21:13:32 <Quibus>   wouterv: I don't actually know which library is used right now to provide that function
jul 10 2013 21:13:40 <Quibus>   i.e. for BiFi 
jul 10 2013 21:13:52 <Quibus>   oh, well, none :) It's not in the header files, so it's probably not supported at all
jul 10 2013 21:14:19 <Quibus>   so, if we would do some kind of link test as well, we could add it to the link flags if it's supported
jul 10 2013 21:38:54 <mth>      Quibus: it might be simpler to just add the lib if the function is probed
jul 10 2013 21:39:10 <mth>      since the situation where the header is present and the lib is not should never occur
jul 10 2013 22:59:10 <mth>      blergh, that clock_gettime thing fits very poorly into the current structure :(
jul 10 2013 23:08:21 <wouterv>  c++11 has new functions to query time .. tomorrow i'll check if those are usabel for us.  If so, then maybe we can avoid this clock_gettime stuff
jul 10 2013 23:08:40 <mth>      ok, that would solve a lot of headaches
jul 10 2013 23:09:16 <mth>      I'll still make notes of them because the build system should be able to handle it, but I prefer to not to have to solve it right now
jul 10 2013 23:21:12 <wouter__> seems like c++11 std::steady_clock is implemented in terms of clock_gettime() or gettimeofday() .. so for gcc/linux using that would be the same as the current implementation
jul 10 2013 23:22:00 <wouter__> it also seems that vc++ already has std::steady_clock, though i can't check the quality of that implementation
jul 10 2013 23:23:10 <wouter__> tomorrow i'll create a patch that switches to the new c++11 stuff, hopefully the mingw link error then disappears

jul 11 2013 22:31:40 <wouterv>  Quibus, mth: I've written a patch to replace clock_gettime with std::chrono::steady_clock() ... it seems to work fine, but unfortunately it only works starting from gcc-4.7
jul 11 2013 22:32:11 <Quibus>   wouterv: aha
jul 11 2013 22:32:24 <Quibus>   wouterv: and is it supported in VC++?
jul 11 2013 22:32:25 <wouterv>  i prefer to keep gcc-4.6 for a while longer, i mean as minimal supported version
jul 11 2013 22:32:55 <Quibus>   My MiNGW is also 4.6
jul 11 2013 22:33:15 <wouterv>  vc++ sort of .. it compiles but it's buggy.  So for vc++ i still kept the SDL implementation (same as today)
jul 11 2013 22:35:39 <Quibus>   so, unfortunately it's a no go for now then...
jul 11 2013 22:35:57 <Quibus>   is it useful to enable it for gcc 4.7 compiles?
jul 11 2013 22:36:10 <wouterv>  i don't know .. what's the current gcc version for mingw?
jul 11 2013 22:36:24 <Quibus>   $ /usr/bin/i686-w64-mingw32-g++ --version
jul 11 2013 22:36:24 <Quibus>   i686-w64-mingw32-g++ (GCC) 4.6.3
jul 11 2013 22:36:24 <mth>      could we fall back to the SDL implementation for GCC 4.6?
jul 11 2013 22:36:48 <wouterv>  if we can't throw away the clock_gettime stuff then there's no advantage yet in switching
jul 11 2013 22:38:21 <mth>      the problem with the SDL implementation is that it's not guaranteed monotonic?
jul 11 2013 22:38:54 <wouterv>  the problem is that the documentation doesn't mentione monotonic or not monotonic at all
jul 11 2013 22:39:52 <wouterv>  the SDL fallback also only has a resolution of 1 millisecond
jul 11 2013 22:40:06 <mth>      so has the clock_gettime() path, no?
jul 11 2013 22:40:32 <mth>      I mean, clock_gettime() itself has a higher res, but its output is converted to milliseconds iirc
jul 11 2013 22:40:47 <wouterv>  the interface offers 1 nanosecond resolution .. though in practice it's likely less
jul 11 2013 22:40:56 <wouterv>  to microseconds
jul 11 2013 22:42:06 <wouterv>  though a very high resolution is probbaly not very important for our purpose
jul 11 2013 22:45:31 <Quibus>   is there another way to solve the link problem on new MinGW?
jul 11 2013 22:45:45 <wouterv>  anyway, to solve the link problems in mingw it's probably ok to also use the SDL fallback
jul 11 2013 22:46:05 <Quibus>   it was already using that on MinGW
jul 11 2013 22:46:14 <Quibus>   because the probe used to be unable to find the function
jul 11 2013 22:46:45 <mth>      yeah, it's probably best to ignore HAVE_CLOCK_GETTIME on MinGW

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2013-10-05 11:55 UTC**
- **status**: open --> closed-fixed
- **assigned_to**: Manuel Bilderbeek

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2013-10-05 11:55 UTC**
Fixed with commit 1588b3c14bd09e583681b90175f804e30e44c57d.

```
Do not compile with clock_gettime on MinGW.

Newer versions of MinGW cross compilers (in Debian) do pass the
HAVE_CLOCK_GETTIME test, but do not implement it in the same library as
what we're using now. When everyone is using this version of MinGW we
should migrate to the library that implements it, but for now, let's
just revert to the SDL method as we did before.
```
